firedrake/ensemble/ensemble_functionspace.py

```diff
@@ -0,0 +1,242 @@
+from firedrake.petsc import PETSc
+from firedrake.functionspace import MixedFunctionSpace
+from ufl.duals import is_primal, is_dual
+from pyop2.mpi import internal_comm
+from functools import cached_property
+from itertools import chain
+
+
+def _is_primal_or_dual(local_spaces, ensemble):
+    """
+    We need to test primal or dual collectively over all ensemble
+    ranks otherwise some may fail when others pass.
+
+    Parameters
+    ----------
+
+    local_spaces : Collection
+        The list of :function:`firedrake.FunctionSpace` on the local ensemble.comm.
+    ensemble : :class:`firedrake.Ensemble`
+        The communicator to test collectively over.
+
+    Returns
+    -------
+    str : Description of the collective space type:
```

Comment:
I would prefer to return an enum: `FunctionSpaceType.PRIMAL` or `FunctionSpaceType.DUAL` (bikeshedding fine)

If strings are already standard then this is fine

Reply:
If this is an enum it should probably go at the UFL level rather than just in one file. I'm ok sticking to strings here, particularly function is supposed to be only an internal implementation detail.